### ISSUE TITLE
Promise support

### ISSUE THREAD

**Issue by ryanseys** [googleapis/google-api-nodejs-client#197]
Suggested by [this tweet](https://twitter.com/rayshan/status/486177042811998208) by @rayshan. Starting the conversation here about promise support.

**Comment by eyy** [comment]
@ryanseys, might I ask to reconsider promise support? Now when we have native promises, and libraries like `co` leveraging generators, writing code like `var calendars = yield calendar.calendarList()` would be great.

Thanks.

**Comment by ryanseys** [comment]
We aren't considering this in the near future but thanks for your feedback. I believe you can wrap the traditional callback structure with promises using other node libraries, so perhaps you could look into that if you need a Promises interface.

**Comment by hess-g** [comment]
Bluebird's promisify works well against this lib. You can see an example [here](https://github.com/danthareja/node-google-apps-script/blob/master/lib/commands/upload.js#L40).

**Comment by LegNeato** [comment]
AFAICT promisify doesn't work when doing JWT auth due to it not matching the node callback pattern. I had to construct a promise manually. I'm using ES7 async/await and arrow functions but should be trivial to do in vanilla ES6:

```js
// Create the promise so we can use await later.
const authPromise = new Promise((resolve, reject) => {
  jwtClient.authorize((err, tokens) => {
    if (err) {
      return reject(err);
    }
    // This is the part that doesn't happen when using promisify.
    jwtClient.setCredentials(tokens);
    resolve(jwtClient);
  });
});

// ...

// Auth as the JWT user. Throws if there is an error.
try {
  await authPromise;
} catch (err) {
  console.log(`ERROR: ${err.message}`);
  throw err;
}
```

**Comment by ronenteva** [comment]
+1 for promise support

**Comment by olalonde** [comment]
On a related note async/await is now implemented in v8: https://news.ycombinator.com/item?id=11710613 which in turn relies on promises... Libraries that don't support a promise based API will quickly run out of fashion.

**Comment by AdamSLevy** [comment]
I think the fact that the JavaScript GAPI Library supports promises while the Node.js GAPI Library doesn't, is a point of confusion for new developers (like myself). Having a consistent JavaScript/Node.js interface would be nice.

**Comment by liviubunda** [comment]
+1 for promise support

**Comment by nhooey** [comment]
Is it possible use BlueBird's [`Promise.fromCallback()`](http://bluebirdjs.com/docs/api/promise.fromcallback.html) on this library?

**Comment by ciscoo** [comment]
@nhooey Yes

```
const { calendar } = require('googleapis')
const Promise = require('bluebird')

const events = calendar('v3').events
const options = {
  auth: '...'
  calendarId: '...',
}

Promise.fromCallback((cb) => events.list(options, cb))
  .then((response) => {
    console.log(response)
  })
  .catch((error) => {
    console.log(error)
  })
```

**Comment by caub** [comment]
Promises are really an improvements over callbacks, above all with async/await now on recent node versions

I hope it will be implemented soon, else a lot of person have to wrap function themselves in promises

**Comment by obiwarn** [comment]
Awaiting await.

Come on Google!

Callbacks are so 2016.

**Comment by ivanseidel** [comment]
why is this closed? Come on Google... its inevitable to evolve to promises...

**Comment by shbmira** [comment]
Google?

**Comment by JustinBeckwith** [comment]
Good news everyone!
![image](https://user-images.githubusercontent.com/534619/38154495-21ad8c3c-3427-11e8-9482-4f900a7e4901.png)

This landed with #1091, and shipped yesterday with v28.  Enjoy :)

**Comment by shbmira** [comment]
@JustinBeckwith, are you kidding me?

There are 2 days left til Apr 1st!

:)

Very good!

🃏
